llvm/include/llvm/BinaryFormat/RootSignatureValidation.h

```diff
@@ -0,0 +1,57 @@
+//===-- llvm/BinaryFormat/RootSignaturesValidation.h ------------*- C++/-*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Validation logic for Root Signatures
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_ROOTSIGNATURE_VALIDATION_H
+#define LLVM_ROOTSIGNATURE_VALIDATION_H
+
+#include "llvm/BinaryFormat/DXContainer.h"
+namespace llvm {
+
+namespace dxbc {
+
+struct RootSignatureValidations {
+
+  static bool isValidRootFlag(uint32_t Flags) { return (Flags & ~0xfff) == 0; }
+
+  static bool isValidVersion(uint32_t Version) {
+    return (Version == 1 || Version == 2);
+  }
+
+  static bool isValidParameterType(dxbc::RootParameterType Type) {
+    switch (Type) {
+    case dxbc::RootParameterType::Constants32Bit:
+      return true;
+    default:
+      return false;
```

Comment:
If this function takes an enum value as a parameter, then we shouldn't have a default here. If we list every possible case, we can rely on the compiler's `-Wcovered-switch` warning to catch errors if we add more enum values and forget to update this switch statement, and if we do need to check for cases that are outside of the range of the enum, then we've already hit UB here. More on this in a later comment.